src/meshes/t8code_mesh.jl

```diff
@@ -0,0 +1,338 @@
+"""
+    T8codeMesh{NDIMS} <: AbstractMesh{NDIMS}
+
+An unstructured curved mesh based on trees that uses the C library 
+['t8code'](https://github.com/DLR-AMR/t8code)
+to manage trees and mesh refinement.
+"""
+mutable struct T8codeMesh{NDIMS, RealT <: Real, IsParallel, NDIMSP2, NNODES} <:
+               AbstractMesh{NDIMS}
+    cmesh       :: Ptr{t8_cmesh} # cpointer to coarse mesh
+    scheme      :: Ptr{t8_eclass_scheme} # cpointer to element scheme
+    forest      :: Ptr{t8_forest} # cpointer to forest
+    is_parallel :: IsParallel
+
+    # This specifies the geometry interpolation for each tree.
+    tree_node_coordinates::Array{RealT, NDIMSP2} # [dimension, i, j, k, tree]
+
+    # Stores the quadrature nodes.
+    nodes::SVector{NNODES, RealT}
+
+    boundary_names   :: Array{Symbol, 2}      # [face direction, tree]
+    current_filename :: String
+
+    ninterfaces :: Int
+    nmortars    :: Int
+    nboundaries :: Int
+
+    function T8codeMesh{NDIMS}(cmesh, scheme, forest, nodes) where {NDIMS}
+        # TODO: Implement MPI parallelization.
+        # if mpi_isparallel()
+        #   if !T8code.uses_mpi()
+        #     error("t8code library does not support MPI")
+        #   end
+        #   is_parallel = Val(true)
+        # else
+        #   is_parallel = Val(false)
+        # end
+        is_parallel = False()
+
+        mesh = new{NDIMS, Float64, typeof(is_parallel), NDIMS + 2, length(nodes)}(cmesh,
+                                                                                  scheme,
+                                                                                  forest,
+                                                                                  is_parallel)
+
+        # Destroy 't8code' structs when the mesh is garbage collected.
+        finalizer(function (mesh::T8codeMesh{NDIMS})
+                      trixi_t8_unref_forest(mesh.forest)
+                  end, mesh)
+
+        return mesh
+    end
+end
+
+function T8codeMesh{NDIMS}(cmesh, scheme, forest, tree_node_coordinates, nodes,
```

Comment:
What's the reason for your choice of writing an inner constructor with a limited amount of fields and an outer constructor with the full set of fields? Why not the other way round?

Reply:
Removed the outer constructor (not actually needed) and moved the intialization code to the inner constructor.